incubator-tools/best-practices/utilities/utilities.py

```diff
@@ -0,0 +1,743 @@
+                        coordinates_xy = bound_poly[0].bounding_poly.normalized_vertices
+                        x1 = [xy.x for xy in coordinates_xy]
+                        y1 = [xy.y for xy in coordinates_xy]
+
+                        try:
+                            page = subentity.page_anchor.page_refs[0].page
+                        except Exception:
+                            page = 0
+                        bbox = [
+                            round(min(x1), 8),
+                            round(min(y1), 8),
+                            round(max(x1), 8),
+                            round(max(y1), 8),
+                        ]
+                        df.loc[len(df.index)] = [
+                            subentity.type_,
+                            subentity.mention_text,
+                            bbox,
+                            page,
+                        ]
+                    except Exception:
+                        continue
+
+            except Exception as e:
+                print(e)
+                continue
+
+            # If no properties were found for the entity, add it to the dataframe
+            if not has_properties:
+                try:
+                    bbox = []
+                    bound_poly = entity.page_anchor.page_refs
+                    coordinates_xy = bound_poly[0].bounding_poly.normalized_vertices
+                    x_1 = [xy.x for xy in coordinates_xy]
+                    y_1 = [xy.y for xy in coordinates_xy]
+                    try:
+                        page = entity.page_anchor.page_refs[0].page
+                    except Exception:
+                        page = 0
+
+                    bbox = [
+                        round(min(x_1), 8),
+                        round(min(y_1), 8),
+                        round(max(x_1), 8),
+                        round(max(y_1), 8),
+                    ]
+                    df.loc[len(df.index)] = [
+                        entity.type_,
+                        entity.mention_text,
+                        bbox,
+                        page,
+                    ]
+                except Exception as e:
+                    print(e)
+                    continue
+
+        return df
+    except Exception as e:
+        print(e)
+        return df
+
+
+def blob_downloader(bucket_name, blob_name):
+    """
+    Downloads a JSON file from a specified Google Cloud Storage bucket
+    and loads it as a Python dictionary.
+
+    Args:
+        bucket_name (str): The name of the GCS bucket from which to download the file.
+        blob_name (str): The name or full path to the JSON blob in the bucket.
+
+    Returns:
+        dict: A dictionary representation of the downloaded JSON file.
+    """
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+    blob = bucket.blob(blob_name)
+
+    contents = blob.download_as_string()
+    return json.loads(contents.decode())
```

Comment:
Is there a need for this function and the `documentai_json_proto_downloader()` function? They seem to do the same thing

Reply:
Blob downloader downloads the file as string while documentai_json_proto_downloader will download the json as document ai object

Reply:
blob_downloader returns the json object and documentai_json_proto_downloader returns the document object

Reply:
Ok, is there a reason that the notebook using `blob_downloader()` doesn't want a Document AI object? Can it use the document ai object? Or is this for reading in different types of JSON data?